File: CS/WinWebSolution.Module/ImportDataListViewController.cs
using System;
using ImportData;
using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp.Templates;

namespace WinWebSolution.Module {
    /// <summary>
    /// Dennis: This ViewController adds the ImportData action and allows you to import data in ListView.
    /// </summary>
    public class ImportDataListViewController : ViewController {
        public const int MaxImportedRecordsCount = 10;
        public const string ActiveKeyImportAction = "ActiveKeyImportAction";
        public const string ActiveKeyImportActionItemRootListViewForPerson = "Item is active only in the root ListView for Person";
        public const string ActiveKeyImportActionItemNestedListViewForPhoneNumber = "Item is active only in nested ListView for PhoneNumber";
        private SingleChoiceAction importDataActionCore = null;

        public ImportDataListViewController() {
            TargetViewType = ViewType.ListView;
            importDataActionCore = CreateImportAction();
        }
        private SingleChoiceAction CreateImportAction() {
            SingleChoiceAction importDataAction = new SingleChoiceAction(this, "ImportData", PredefinedCategory.RecordEdit);
            importDataAction.Execute += importDataAction_Execute;
            ChoiceActionItem item1 = new ChoiceActionItem();
            ChoiceActionItem item2 = new ChoiceActionItem();
            importDataAction.Caption = "Import Data";
            importDataAction.ImageName = "Attention";
            item1.Caption = "In root ListView";
            item1.Data = "Person_ListView";
            item2.Caption = "In nested ListView";
            item2.Data = "Party_PhoneNumbers_ListView";
            importDataAction.Items.Add(item1);
            importDataAction.Items.Add(item2);
            importDataAction.PaintStyle = ActionItemPaintStyle.CaptionAndImage;
            importDataAction.ItemType = SingleChoiceActionItemType.ItemIsOperation;
            return importDataAction;
        }
        protected virtual void ImportData(SingleChoiceActionExecuteEventArgs e) {
            ChoiceActionItem activeItem = e.SelectedChoiceActionItem;
            ListView lv = (ListView)View;
            ImportDataManager importManager = new ImportDataManager(Application);
            switch (activeItem.Data.ToString()) {
                case "Person_ListView":
                    importManager.ImportData<Person>(MaxImportedRecordsCount, ImportDataLogic.CreateCoolPersonImportDataFromXmlFileDelegate(), null, lv, true);
                    //Dennis: This line won't be executed unless you handle the exception thrown from the previus ImportData call.
                    //ImportDataManager.ImportData<Person>(MaxImportedRecordsCount, ImportDataLogic.CreateDummyPersonImportDataDelegate(), ImportDataLogic.CreateDummyPersonValidateDataDelegate(), lv, true);
                    break;
                case "Party_PhoneNumbers_ListView":
                    PropertyCollectionSource pcs = lv.CollectionSource as PropertyCollectionSource;
                    if (pcs != null) {
                        importManager.ImportData<PhoneNumber>(MaxImportedRecordsCount, ImportDataLogic.CreateDummyPhoneNumberImportDataDelegate(), ImportDataLogic.CreateDummyPhoneNumberValidateDataDelegate(), lv, true);
                    }
                    break;
            }
        }
        private void importDataAction_Execute(object sender, SingleChoiceActionExecuteEventArgs e) {
            ImportData(e);
        }
        protected override void UpdateActionActivity(ActionBase action) {
            base.UpdateActionActivity(action);
            bool rootListViewForPersonCondition = View.IsRoot && View is ListView && View.ObjectTypeInfo.Type == typeof(Person);
            bool nestedListViewForPhoneNumberCondition = !View.IsRoot && View is ListView && View.ObjectTypeInfo.Type == typeof(PhoneNumber);
            ImportDataAction.Active[ActiveKeyImportAction] = rootListViewForPersonCondition || nestedListViewForPhoneNumberCondition;
            ImportDataAction.Items[0].Active[ActiveKeyImportActionItemRootListViewForPerson] = rootListViewForPersonCondition;
            ImportDataAction.Items[1].Active[ActiveKeyImportActionItemNestedListViewForPhoneNumber] = nestedListViewForPhoneNumberCondition;
        }
        public SingleChoiceAction ImportDataAction {
            get {
                return importDataActionCore;
            }
        }
    }
}

File: VB/WinWebSolution.Module/ImportDataListViewController.vb
Imports System
Imports ImportData
Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp.Templates

Namespace WinWebSolution.Module
	''' <summary>
	''' Dennis: This ViewController adds the ImportData action and allows you to import data in ListView.
	''' </summary>
	Public Class ImportDataListViewController
		Inherits ViewController

		Public Const MaxImportedRecordsCount As Integer = 10
		Public Const ActiveKeyImportAction As String = "ActiveKeyImportAction"
		Public Const ActiveKeyImportActionItemRootListViewForPerson As String = "Item is active only in the root ListView for Person"
		Public Const ActiveKeyImportActionItemNestedListViewForPhoneNumber As String = "Item is active only in nested ListView for PhoneNumber"
		Private importDataActionCore As SingleChoiceAction = Nothing

		Public Sub New()
			TargetViewType = ViewType.ListView
			importDataActionCore = CreateImportAction()
		End Sub
		Private Function CreateImportAction() As SingleChoiceAction
'INSTANT VB NOTE: The variable importDataAction was renamed since Visual Basic does not handle local variables named the same as class members well:
			Dim importDataAction_Renamed As New SingleChoiceAction(Me, "ImportData", PredefinedCategory.RecordEdit)
			AddHandler importDataAction_Renamed.Execute, AddressOf importDataAction_Execute
			Dim item1 As New ChoiceActionItem()
			Dim item2 As New ChoiceActionItem()
			importDataAction_Renamed.Caption = "Import Data"
			importDataAction_Renamed.ImageName = "Attention"
			item1.Caption = "In root ListView"
			item1.Data = "Person_ListView"
			item2.Caption = "In nested ListView"
			item2.Data = "Party_PhoneNumbers_ListView"
			importDataAction_Renamed.Items.Add(item1)
			importDataAction_Renamed.Items.Add(item2)
			importDataAction_Renamed.PaintStyle = ActionItemPaintStyle.CaptionAndImage
			importDataAction_Renamed.ItemType = SingleChoiceActionItemType.ItemIsOperation
			Return importDataAction_Renamed
		End Function
		Protected Overridable Sub ImportData(ByVal e As SingleChoiceActionExecuteEventArgs)
			Dim activeItem As ChoiceActionItem = e.SelectedChoiceActionItem
			Dim lv As ListView = CType(View, ListView)
			Dim importManager As New ImportDataManager(Application)
			Select Case activeItem.Data.ToString()
				Case "Person_ListView"
					importManager.ImportData(Of Person)(MaxImportedRecordsCount, ImportDataLogic.CreateCoolPersonImportDataFromXmlFileDelegate(), Nothing, lv, True)
					'Dennis: This line won't be executed unless you handle the exception thrown from the previus ImportData call.
					'ImportDataManager.ImportData<Person>(MaxImportedRecordsCount, ImportDataLogic.CreateDummyPersonImportDataDelegate(), ImportDataLogic.CreateDummyPersonValidateDataDelegate(), lv, true);
				Case "Party_PhoneNumbers_ListView"
					Dim pcs As PropertyCollectionSource = TryCast(lv.CollectionSource, PropertyCollectionSource)
					If pcs IsNot Nothing Then
						importManager.ImportData(Of PhoneNumber)(MaxImportedRecordsCount, ImportDataLogic.CreateDummyPhoneNumberImportDataDelegate(), ImportDataLogic.CreateDummyPhoneNumberValidateDataDelegate(), lv, True)
					End If
			End Select
		End Sub
		Private Sub importDataAction_Execute(ByVal sender As Object, ByVal e As SingleChoiceActionExecuteEventArgs)
			ImportData(e)
		End Sub
		Protected Overrides Sub UpdateActionActivity(ByVal action As ActionBase)
			MyBase.UpdateActionActivity(action)
			Dim rootListViewForPersonCondition As Boolean = View.IsRoot AndAlso TypeOf View Is ListView AndAlso View.ObjectTypeInfo.Type Is GetType(Person)
			Dim nestedListViewForPhoneNumberCondition As Boolean = (Not View.IsRoot) AndAlso TypeOf View Is ListView AndAlso View.ObjectTypeInfo.Type Is GetType(PhoneNumber)
			ImportDataAction.Active(ActiveKeyImportAction) = rootListViewForPersonCondition OrElse nestedListViewForPhoneNumberCondition
			ImportDataAction.Items(0).Active(ActiveKeyImportActionItemRootListViewForPerson) = rootListViewForPersonCondition
			ImportDataAction.Items(1).Active(ActiveKeyImportActionItemNestedListViewForPhoneNumber) = nestedListViewForPhoneNumberCondition
		End Sub
		Public ReadOnly Property ImportDataAction() As SingleChoiceAction
			Get
				Return importDataActionCore
			End Get
		End Property
	End Class
End Namespace